FLT/Deformations/Lemmas.lean

```diff
@@ -0,0 +1,168 @@
+  smul a f := ⟨a • f, .const_smul f.2 _⟩
+  one_smul _ := by ext; exact one_smul _ _
+  mul_smul a b f := by ext x; show (a * b) • (f x) = a • b • (f x); simp [mul_smul]
+
+@[simp]
+lemma ContinuousMonoidHom.coe_smul
+    {α M G : Type*} [Monoid M] [Monoid G] [Monoid α] [MulDistribMulAction α G]
+    [TopologicalSpace M] [TopologicalSpace G] [ContinuousConstSMul α G]
+    (a : α) (f : M →ₜ* G) :
+    ⇑(a • f) = a • ⇑f := rfl
+
+instance {G : Type*} [Group G] [TopologicalSpace G] [ContinuousMul G] :
+    ContinuousConstSMul (ConjAct G) G :=
+  ⟨fun _ ↦ .mul (continuous_const.mul continuous_id) continuous_const⟩
+
+@[to_additive]
+instance IsTopologicalGroup.discreteUniformity
+    {G : Type*} [Group G] [TopologicalSpace G] [IsTopologicalGroup G] [DiscreteTopology G] :
+    letI := IsTopologicalGroup.toUniformSpace G
+    DiscreteUniformity G := by
+  simp only [discreteUniformity_iff_idRel_mem_uniformity, uniformity_eq_comap_nhds_zero',
+    nhds_discrete, Filter.pure_zero, Filter.mem_comap, Filter.mem_zero]
+  exact ⟨{1}, by simp [Set.subset_def, div_eq_one]⟩
+
+lemma IsLocalRing.map_maximalIdeal {R S} [CommRing R] [CommRing S]
+    [IsLocalRing R] [IsLocalRing S] (f : R →+* S) (hf : Function.Surjective f) :
+    (maximalIdeal R).map f = maximalIdeal S := by
+  have := (IsLocalRing.local_hom_TFAE f).out 0 4
+  rw [← this.mp (by exact .of_surjective f hf), Ideal.map_comap_of_surjective f hf]
+
+lemma IsLocalRing.ResidueField.map_surjective {R S : Type*} [CommRing R] [CommRing S]
+    [IsLocalRing R] [IsLocalRing S] (f : R →+* S) [IsLocalHom f] (H : Function.Surjective f) :
+    Function.Surjective (IsLocalRing.ResidueField.map f) :=
+  Ideal.Quotient.lift_surjective_of_surjective _ _ (Ideal.Quotient.mk_surjective.comp H)
+
+open Topology in
+@[to_additive]
+lemma MonoidHom.continuous_iff_isOpen_ker {α β : Type*} [TopologicalSpace α] [TopologicalSpace β]
+    [Group α] [MulOneClass β] {f : α →* β} [DiscreteTopology β] [IsTopologicalGroup α] :
+    Continuous f ↔ IsOpen (X := α) f.ker := by
+  refine ⟨fun H ↦ H.1 {1} (isOpen_discrete _), fun H ↦ continuous_of_continuousAt_one _ ?_⟩
+  simpa [ContinuousAt] using Filter.eventually_of_mem (H.mem_nhds (by simp)) (by simp)
+
+lemma RingHom.continuous_iff_isOpen_ker {α β : Type*} [TopologicalSpace α] [TopologicalSpace β]
+    [Ring α] [Semiring β] {f : α →+* β} [DiscreteTopology β] [IsTopologicalAddGroup α] :
+    Continuous f ↔ IsOpen (X := α) (RingHom.ker f) :=
+  AddMonoidHom.continuous_iff_isOpen_ker (f := f.toAddMonoidHom)
+
+open Topology in
+lemma continuousAt_discrete_rng {α β : Type*} [TopologicalSpace α] [TopologicalSpace β]
+    {f : α → β} {x : α} [DiscreteTopology β] :
+    ContinuousAt f x ↔ ∀ᶠ y in 𝓝 x, f y = f x := by
+  simp [ContinuousAt]
+
+nonrec
+instance ContinuousAlgHom.isLocalHom_id {R A : Type*}
+    [CommSemiring R] [Semiring A] [Algebra R A] [TopologicalSpace A] :
+    IsLocalHom (ContinuousAlgHom.id R A) := by
+  convert isLocalHom_id A
+  exact ⟨fun ⟨H⟩ ↦ ⟨H⟩, fun ⟨H⟩ ↦ ⟨H⟩⟩
```

Comment:
To avoid this `⟨fun ⟨H⟩ ↦ ⟨H⟩, fun ⟨H⟩ ↦ ⟨H⟩⟩` boilerplate, you probably want to add `@[mk_iff]` to `IsLocalHom`, and the defeq should work after rewriting by `isLocalHom_iff`.